pkg/kubelet/pod_container_deletor.go

```diff
@@ -80,20 +81,27 @@ func getContainersToDeleteInPod(filterContainerId string, podStatus *kubecontain
 
 	// Find the exited containers whose name matches the name of the container with id being filterContainerId
 	var candidates containerStatusbyCreatedList
+	var infraCandidates containerStatusbyCreatedList
 	for _, containerStatus := range podStatus.ContainerStatuses {
 		if containerStatus.State != kubecontainer.ContainerStateExited {
 			continue
 		}
 		if matchedContainer == nil || matchedContainer.Name == containerStatus.Name {
-			candidates = append(candidates, containerStatus)
+			if containerStatus.Name == leaky.PodInfraContainerName {
+				infraCandidates = append(infraCandidates, containerStatus)
+			} else {
+				candidates = append(candidates, containerStatus)
+			}
 		}
 	}
 
 	if len(candidates) <= containersToKeep {
-		return containerStatusbyCreatedList{}
+		// Always destroy infra containers to ensure networking gets torn down
```

Comment:
@dcbw I'd like to try making the change small and localized since it's so close to the release (unless this is not targeted for 1.5).

Change in this file is very docker-specific even though it's used by all implementations (rkt, cri, etc). On top of that, kubenet has its own fix that handles the case where the infra containers get removed underneath kubelet. By leaving everything to the container GC, you have already fixed the leaks in most situation (e.g., docker restart). Would that be sufficient?